specreduce/extract.py

```diff
@@ -143,3 +145,108 @@ def _ap_weight_image(trace, width, disp_axis, crossdisp_axis, image_shape):
+    """
+
+    def __call__(self, image, variance, trace_object,
+                 disp_axis=1, crossdisp_axis=0,
+                 bkgrd_prof=models.Polynomial1D(2)):
+        """
+        Run the Horne calculation on a region of an image and extract a
+        1D spectrum.
+
+        Parameters
+        ----------
+
+        image : `~astropy.nddata.CCDData` or array-like, required
+            The input 2D spectrum from which to extract a source.
+
+        variance : `~astropy.nddata.CCDData` or array-like, required
+            The associated variances for each pixel in the image.
+            Should have the same dimensions as `image`.
+
+        trace_object : `~specreduce.tracing.Trace`, required
+            The associated 1D trace object created for the 2D image.
+
+        disp_axis : int, optional
+            The index of the image's dispersion axis. [default: 1]
+
+        crossdisp_axis : int, optional
+            The index of the image's cross-dispersion axis. [default: 0]
+
+        bkgrd_prof : `~astropy.modeling.Model`
+            A model for the image's background flux.
+            [default: models.Polynomial1D(2)]
+
+        Returns
+        -------
+        spec_1d : `~specutils.Spectrum1D`
+            The final, Horne extracted 1D spectrum.
+        """
+        # co-add signal in each column
+        ncols = image.shape[crossdisp_axis]
+        xd_pixels = np.arange(ncols) # y plot dir / x spec dir
+        coadd = image.sum(axis=disp_axis) / ncols
+
+        # fit source profile, using Gaussian model as a template
+        # NOTE: could add argument for users to provide their own model
+        gauss_prof = models.Gaussian1D(amplitude=coadd.max(),
+                                       mean=coadd.argmax(), stddev=2)
+
+        # Fit extraction kernel to column with combined gaussian/bkgrd model
+        ext_prof = gauss_prof + bkgrd_prof
+        fitter = fitting.LevMarLSQFitter()
+        fit_ext_kernel = fitter(ext_prof, xd_pixels, coadd)
+
+        # generate 1D spectrum, column by column
+        extracted = []
+        for col_pix in range(image.shape[disp_axis]):
+            # set gaussian model's mean as column's corresponding trace value
+            fit_ext_kernel.mean_0 = trace_object.trace[col_pix]
+            # kernel_col.stddev_0 = self.fwhm_fit(x)
+            # NOTE: support for variable FWHMs forthcoming
+
+            # fit compound model to column
+            kernel_vals = fit_ext_kernel(xd_pixels)
+            # NOTE: support user-provided kernels? would require different handling
+            # NOTE: allow smoothing in x direction via boxcar/x axis collapse/etc?
+
+            # fetch matching columns from original and variance images
+            image_col = image[:, col_pix]
+            variance_col = variance[:, col_pix]
+
+            # calculate kernel normalization
+            g_x = np.ma.sum(kernel_vals**2 / variance_col)
+            if np.ma.is_masked(g_x):
+                continue
```

Comment:
when the `continue` is triggered, won't the length of `pixels` and `extracted` be different when you create the `Spectrum1D`?

Reply:
@kecnry you're right. I was switching between approaches of appending to an empty list and filling in a pre-made array of size `pixels`. I'm still thinking about how both fare with vectorization.

Reply:
i think using `np.fromfunction` is the approach we ultimately want to use here to leverage numpy broadcasting as much as possible. it'll require some refactoring/rearranging of the logic here so that might be something to leave for a future optimization. `np.fromiter` is another option that's more efficient than using lists.

Reply:
that all said, it's also fine to have the initial implementation be straightforward and not optimal performance. provides a baseline for future improvements to compare to.

Reply:
I got rid of the `continue` in the latest commit. Since the arrays are masked, any NaNs can just be carried on to the final, extracted 1D spectrum.

Reply:
@tepickering, I didn't find a sensible way to build `kernel_vals` or `norms` with `np.fromiter` or `np.fromfunction`, so I still create them with a loop. Continuing to append to a list was a small percentage faster than filling arrays.

`kernel_vals` becomes a 2D array that needs the fit object's mean attribute in each column to be modified before any fit results are stored in it. If that pre-modification wasn't necessary, I could see how `np.fromfunction` would be useful, plus the array of index locations it passes to the inner function would eliminate the need for `xd_pixels`. `np.fromiter` can only return 1D arrays, so it seems like a no-go here.

Each entry in `norms` must be calculated immediately after the corresponding fit in `kernel_vals`. This means we'd need to get `norms` from the same call to either `numpy` function as `kernel_vals`, and I don't think either is set up to give two separate arrays as output.

An alternate approach I tried for `norms` was to turn `fit_ext_kernel` into an array of separate fit objects for each column. I could then call `np.fromiter` twice on an inner function that used `getattr()` -- one call for amplitude and the other for standard deviation. However, even creating the array of fit objects was much slower than using the original loop.